src/include/api.h

```diff
@@ -32,30 +32,37 @@
 #define WT_SINGLE_THREAD_CHECK_STOP(s)
 #endif
 
-#define API_SESSION_PUSH(s, h, n, dh)       \
-    WT_DATA_HANDLE *__olddh = (s)->dhandle; \
-    const char *__oldname = (s)->name;      \
-    (s)->dhandle = (dh);                    \
+#define API_SESSION_PUSH(s, h, n, dh)                                                        \
+    WT_DATA_HANDLE *__olddh = (s)->dhandle;                                                  \
+    const char *__oldname;                                                                   \
+    /* If this isn't an API reentry, the name should be NULL and the counter should be 0. */ \
+    WT_ASSERT(session, (s)->name != NULL || s->api_call_counter == 0);                       \
```

Comment:
Checking whether the reference counting is correct by comparing it with the api call name.

Reply:
Nice one. I had to think hard about whether this assertion was OK, since on first glance it ignores the " No code before this line, otherwise error handling won't be correct." comment that precedes it.

I'm pretty sure that references usage of __wt_err and friends, not assertions so I think it's OK.